twilight-gateway-queue/src/in_memory.rs

```diff
@@ -0,0 +1,219 @@
+//! Memory based [`Queue`] implementation and supporting items.
+
+use super::{Queue, IDENTIFY_INTERVAL};
+use std::{collections::VecDeque, fmt::Debug, iter};
+use tokio::{
+    sync::{mpsc, oneshot},
+    task::yield_now,
+    time::{sleep_until, Duration, Instant},
+};
+
+/// Possible messages from the [`InMemoryQueue`] to the [`runner`].
+#[derive(Debug)]
+enum Message {
+    /// Request a permit.
+    Request {
+        /// For this shard.
+        shard: u64,
+        /// Indicate readiness through this sender.
+        tx: oneshot::Sender<()>,
+    },
+    /// Update the runner's settings.
+    Update(Settings),
+}
+
+/// [`runner`]'s settings.
+#[derive(Debug)]
+struct Settings {
+    /// The maximum number of concurrent permits to grant. `0` instantly grants
+    /// all permits.
```

Comment:
I think this is a bit too hidden considering it replaces `NoOpQueue` entirely. We should consider mentioning at module level

Reply:
I agree but I had trouble finding an appropriate spot for it. Since `NoOpQueue` is mostly relevant for twilight-gateway I think it would actually be better to mention it in `ConfigBuilder::queue` instead.